ocamlnat: fixes

trefis opened the pull request:
The build was broken, again.
If it breaks again I probably won't do another PR here and just wait for flambda to get merged ( https://github.com/chambart/ocaml-1/pull/21 is a working version of ocamlnat on flambda supporting both backends).

With the build fixed I noticed that https://github.com/ocaml/ocaml/commit/9216c69017dedfe0411851bf8218d6c71c5019c9 didn't handle the case were an unit was generated by the toplevel. I think my patch is correct but would appreciate it if @lpw25 could have a look.

trefis committed:
ocamlnat: fix compilation

gasche commented on toplevel/opttoploop.ml:
```diff
@@ -165,7 +165,8 @@ let load_lambda ppf (size, lam) =
   in
   let fn = Filename.chop_extension dll in
-  Asmgen.compile_implementation ~toplevel:need_symbol fn ppf (size, slam);
+  Asmgen.compile_implementation ~sourcefile:"" ~toplevel:need_symbol fn ppf
```
``` ocaml
let dummy_sourcefile = "native toplevel"

...
  Asm.compile_implementation ~sourcefile:dummy_sourcefile 
```

gasche replied:
Thinking more about it, this would justify changing the `compile_implentation` from `sourcefile` to `source_provenance`. @chambart, is there a good reason for passing `~sourcefile:"pack"` [in Asmpackager](https://github.com/trefis/ocaml/blob/ocamlnat/asmcomp/asmpackager.ml#L94)?

gasche replied:
You forgot to change this callsite, right?

trefis replied:
Well spotted, it must have been lost during the rebase.

gasche commented on toplevel/opttoploop.ml:
```diff
@@ -217,7 +218,8 @@ let execute_phrase print_outcome ppf phr =
       incr phrase_seqid;
       phrase_name := Printf.sprintf "TOP%i" !phrase_seqid;
-      Compilenv.reset ?packname:None !phrase_name;
+      Compilenv.reset ~source_provenance:Timings.Startup
```
You could add a `Toplevel` case to `Timings.source_provenance` datatype?

trefis replied:
I actually left a comment for @mshinwell or @chambart about that on my other PR.
I hope to get some feedback from them on this.

gasche commented:
It's not clear to me while the `not (Env.is_imported modname)` thing is needed (it's a bit ugly so naturally a bit suspicious), given that the toplevel already has a [separate table](https://github.com/trefis/ocaml/blob/ocamlnat/asmcomp/compilenv.ml#L197-L198) to record approximations and this table [gets queried first](https://github.com/trefis/ocaml/blob/ocamlnat/asmcomp/compilenv.ml#L202) when approximations are queried. Do you understand when `get_global_info` gets called on a toplevel-compiled phrase?

trefis commented:
> Do you understand when get_global_info gets called on a toplevel-compiled phrase?

Yes, it's called by `symbol_for_global`, itself called from `Closure`.

gasche commented:
Then maybe you could try adding a lookup to `toplevel_approx` in `symbol_for_global`, so that its code structure matches the one of `global_approx`?

trefis commented:
I considered that but found it uglier than my current solution. But you might be right (also, your way might be better, performance wise).
I'll let @lpw25 a chance to voice his opinion on the matter and if he doesn't I'll just update the code in a few days as per your comment.

lpw25 commented:
The `is_imported` fix is not quite safe (at least not with flambda) as it is possible to want the .cmx of a file which was not imported directly. If module `Foo` uses module `Bar` in its implementation but not in its interface then modules using `Foo` may need to load the cmx of `Bar` even though `Bar` is not registered as an imported module. This doesn't matter for `is_imported_opaque` since obviously this can't happen if `Bar` is opaque.

So I would go with @gasche's solution, which is probably better anyway as it might allow the inlining of values from toplevel units.

trefis committed:
ocamlnat: silence warning 58

trefis committed:
timings: add a Toplevel source_provenance

trefis commented:
Thanks for the review!
I updated the commit to use @gasche's solution, and after discussing it with @mshinwell I added a `Toplevel` case to `source_provenance`.

gasche commented:
My only pending comment is a suggestion to change `compile_implementation` to take a `~source_provenance` parameter rather than `~sourcefile`. What do you think? This is a non-local change so maybe a separate commit would be reasonable, but I think you could do it -- then I'd merge the PR.

(This raises the question of whether some _other_ `~sourcefile`-using functions should be converted to the more general `~source_provenance` to be robust to future changes.)

trefis committed:
use Timings.source_provenance in more places

trefis commented:
Done.

gasche commented:
Sorry to be a bore, but could you squash the three ocamlnat-specific commits together? Currently the commit message for 4835a1317457f73d26f45d0667463d468b2dc69f is wrong.
